File: src/com/kiibou/Map.java
package com.kiibou;


import space.kiibou.GApplet;
import space.kiibou.gui.GraphicsElement;
import space.kiibou.gui.Grid;
import space.kiibou.gui.VerticalList;

import java.util.function.Consumer;

public class Map extends GraphicsElement {

    private final int tilesX;
    private final int tilesY;
    private final int bombs;
    private Grid<Tile> tiles;
    private VerticalList verticalList;
    private ControlBar controlBar;

    public Map(final GApplet app, final int x, final int y, final int tilesX, final int tilesY, final int scale, final int bombs) {
        super(app, x, y, tilesX * Tile.tileHeight * scale, tilesY * Tile.tileHeight * scale, scale);
        this.tilesX = tilesX;
        this.tilesY = tilesY;
        this.bombs = bombs;
    }

    @Override
    public void preInitImpl() {
        verticalList = new VerticalList(getApp(), getX(), getY(), getScale());
        addChild(verticalList);

        controlBar = new ControlBar(getApp(), getScale(), this);
        verticalList.addChild(controlBar);

        tiles = createGrid(tilesX, tilesY);
        verticalList.addChild(tiles);
    }

    @Override
    public void initImpl() {
        controlBar.setWidth(verticalList.getChild(0).getInnerWidth());
        resize(verticalList.getWidth(), verticalList.getHeight());
    }

    @Override
    public void postInitImpl() {

    }

    @Override
    public void drawImpl() {

    }

    private Grid<Tile> createGrid(final int tilesX, final int tilesY) {
        final Grid<Tile> tiles = new Grid<>(getApp(), 0, 0, tilesX, tilesY, getScale());

        for (int x = 0; x < tilesX; x++) {
            for (int y = 0; y < tilesY; y++) {
                final Tile tile = new Tile(getApp(), this, getScale(), x, y);
                tiles.put(x, y, tile);
            }
        }

        return tiles;
    }

    public void revealTile(int x, int y, TileType type) {
        tiles.get(x, y).setType(type);
        tiles.get(x, y).setRevealed(true);
    }

    public void win() {
        controlBar.setSmiley(SmileyStatus.GLASSES);

        forEachTile(Tile::deactivate);
    }

    public void loose() {
        forEachTile(Tile::deactivate);

        controlBar.setSmiley(SmileyStatus.DEAD);
    }

    public void restart() {
        forEachTile(Tile::reset);
        controlBar.setSmiley(SmileyStatus.NORMAL);
        controlBar.getBombsLeft().setValue(bombs);
    }

    public void tileFlag(final int x, final int y, final boolean flagged) {
        if (flagged) {
            controlBar.getBombsLeft().dec();
        } else {
            controlBar.getBombsLeft().inc();
        }

        tiles.get(x, y).setFlagged(flagged);
    }

    private void forEachTile(Consumer<Tile> action) {
        tiles.forEach(action);
    }

    public int getBombs() {
        return bombs;
    }

    public ControlBar getControlBar() {
        return controlBar;
    }
}

File: src/com/kiibou/Map.kt
package com.kiibou

import space.kiibou.GApplet
import space.kiibou.gui.*

class Map(app: GApplet, private val tilesX: Int, private val tilesY: Int, val bombs: Int)
    : GraphicsElement(app) {

    private val margin = tileWidth / 4
    private val marginProp = scaleProp.multiply(margin)

    private val box = BorderBox(app).also {
        it.borderStyle = BorderStyle.OUT
        addChild(it)
    }

    private val verticalList = VerticalList(app, margin).also {
        box.addChild(it)
        it.xProp.bind(box.xProp.add(box.borderWidthProp).add(marginProp))
        it.yProp.bind(box.yProp.add(box.borderHeightProp).add(marginProp))
        box.innerWidthProp.bind(it.widthProp.add(marginProp.multiply(2)))
        box.innerHeightProp.bind(it.heightProp.add(marginProp.multiply(2)))
    }

    private val tiles = Grid<Tile>(app, tilesX, tilesY).also {
        (0 until tilesX).forEach { x ->
            (0 until tilesY).forEach { y ->
                it[x, y] = Tile(app, this, x, y)
            }
        }
    }

    private val tilesBox = BorderBox(app).also {
        it.addChild(tiles)
        it.bindProps(tiles)
    }

    val controlBar = ControlBar(app, margin, this).also {
        it.widthProp.bind(tiles.widthProp)
    }

    private val controlBarBox = BorderBox(app).also {
        it.addChild(controlBar)
        it.bindProps(controlBar)
    }

    init {
        verticalList += controlBarBox
        verticalList += tilesBox

        widthProp.bind(box.widthProp)
        heightProp.bind(box.heightProp)
    }

    public override fun preInitImpl() {}
    public override fun initImpl() {
        (app as Minesweeper).client.sendJson(mapper.valueToTree(MapInfo(tilesX, tilesY, bombs)))
    }

    public override fun postInitImpl() {}
    public override fun drawImpl() {}

    fun revealTile(x: Int, y: Int, type: TileType) {
        tiles[x, y]!!.type = type
        tiles[x, y]!!.revealed = true
    }

    fun revealTiles(revealTiles: RevealTiles) {
        revealTiles.tiles.forEach {
            revealTile(it.x, it.y, TileType.getTypeFromValue(it.type))
        }
    }

    fun win() {
        controlBar.setSmiley(SmileyStatus.GLASSES)
        forEachTile(Tile::deactivate)
    }

    fun loose() {
        forEachTile(Tile::deactivate)
        controlBar.setSmiley(SmileyStatus.DEAD)
    }

    fun restart() {
        forEachTile(Tile::reset)
        controlBar.setSmiley(SmileyStatus.NORMAL)
        controlBar.bombsLeft.value = bombs
    }

    fun tileFlag(flagInfo: FlagInfo) {
        when (flagInfo.toggle) {
            true -> controlBar.bombsLeft.dec()
            false -> controlBar.bombsLeft.inc()
        }

        tiles[flagInfo.x, flagInfo.y]!!.flagged = flagInfo.toggle
    }

    private inline fun forEachTile(action: (Tile) -> Unit) {
        tiles.forEach(action)
    }

}